src/module.c

```diff
@@ -109,54 +110,118 @@ English.
 Returns OK on success, NOADD if the document was not added due to an IF expression not evaluating to
 true or an error if something went wrong.
 */
-static int doAddDocument(RedisModuleCtx *ctx, RedisModuleString **argv, int argc, int canBlock) {
-  int nosave = RMUtil_ArgExists("NOSAVE", argv, argc, 1);
-  int fieldsIdx = RMUtil_ArgExists("FIELDS", argv, argc, 1);
-  int replace = RMUtil_ArgExists("REPLACE", argv, argc, 1);
-  int partial = RMUtil_ArgExists("PARTIAL", argv, argc, 1);
 
-  if (canBlock) {
-    canBlock = CheckConcurrentSupport(ctx);
+static int parseDocumentOptions(AddDocumentOptions *opts, ArgsCursor *ac, QueryError *status) {
+  // Assume argc and argv are at proper indices
+  int nosave = 0, replace = 0, partial = 0, foundFields = 0;
+  opts->fieldsArray = NULL;
+  opts->numFieldElems = 0;
+  opts->options = 0;
+
+  ACArgSpec specs[] = {{.name = "NOSAVE", .type = AC_ARGTYPE_BOOLFLAG, .target = &nosave},
```

Comment:
specs is not a good name here cause we use spec to the actual spec struct (IndexSpec), pls change it (maybe specsArgs?)

Reply:
We should really rename IndexSpec to something else.. I'll file an issue.

In this function there isn't a whole lot of confusion because there's no actual IndexSpec here, or anything really stateful (outside of the options themselves), but I'll change it if it's easier to read like that.

Reply:
you did not changed the name ..